### ISSUE TITLE
关于AVT数据集，为什么我可视化出来的数据不是论文中这种血管树？？？？？？

### ISSUE THREAD

**Issue by sansuitaibai** [HongkunSun/ParaTransCNN#13]
可视化AVT，并不是论文那种啊？？数据集给错了？？？

**Comment by HongkunSun** [comment]
可视化的是标签还是模型推理出来的结果？数据不会出错的

**Comment by sansuitaibai** [comment]
可视化的是标签，R，K，D三种可视化出来， 难道是我可视化代码写错了？




spw
***@***.***



&nbsp;



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "HongkunSun/ParaTransCNN"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月21日(星期三) 下午5:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [HongkunSun/ParaTransCNN] 关于AVT数据集，为什么我可视化出来的数据不是论文中这种血管树？？？？？？ (Issue #13)



 HongkunSun left a comment (HongkunSun/ParaTransCNN#13)
 
可视化的是标签还是模型推理出来的结果？数据不会出错的

**Comment by HongkunSun** [comment]
可视化使用3D Slicer

**Comment by sansuitaibai** [comment]
这是我根据作者你提供代码得到的我的模型结果，难道是我的可视化错了吗？AVT不是一个2分类的问题吗？可视化时



spw
***@***.***






下面是我的可视化代码：
def calculate_metric_percase(pred, gt):
    pred[pred &gt; 0] = 1
    gt[gt &gt; 0] = 1
    if pred.sum() &gt; 0 and gt.sum()&gt;0:
        dice = metric.binary.dc(pred, gt)
        hd95 = metric.binary.hd95(pred, gt)
        return dice, hd95
    elif pred.sum() &gt; 0 and gt.sum()==0:
        return 1, 0
    else:
        return 0, 0

import matplotlib.pyplot as plt
import matplotlib
matplotlib.use("Agg")
def visualize_segmentation(image, label, pred, cam_hotmap_list, cam_hotmap_name,
                           save_path, slice_name, model_name):
    # 定义类别颜色映射
    category_colors = {
        1: (255, 0, 0),      # 红色 (AVT)
    }
    # 归一化原始图像到 0-255 范围
    image = (image - image.min()) / (image.max() - image.min())
    image = (image * 255).astype(np.uint8)

    # 创建一个彩色图像，初始值是原始灰度图
    colored_image = np.stack([image] * 3, axis=-1)  # (H, W, 3)
    colored_image_pre = np.stack([image] * 3, axis=-1)  # (H, W, 3)

    # 替换类别区域的颜色，但背景保留原图
    for class_idx, color in category_colors.items():
        mask = (label == class_idx)
        colored_image[mask] = np.array(color, dtype=np.uint8)
        mask_pre = (pred == class_idx)
        colored_image_pre[mask_pre] = np.array(color, dtype=np.uint8)
    # 绘制可视化结果
    img_num = 1
    fig, axes = plt.subplots(1, 1, figsize=(5 * img_num, 8))
    # 右侧显示填充颜色后的图像
    axes.imshow(colored_image)
    # axes[1].set_title(f"{model_name}_Label: ({slice_name})")
    axes.axis("off")
    plt.tight_layout()
    # 保存图片
    if save_path is not None:
        plt.savefig(save_path, dpi=300, bbox_inches='tight')
        print(f"图片保存成功: {save_path}")


#draw slice
def draw_slice(net,images, labels, preds, case_name, patch_size=None):
    net.eval()
    print(f'images: {images.shape}')
    print(f'labels: {labels.shape}')
    print(f'preds: {preds.shape}')
    print(f'case_name: {case_name}')
    num, h, w = images.shape
    for index in range(num):
        img = images[index]
        lab = labels[index]
        if not np.any(lab):
            continue
        pred = preds[index]

        save_path = "./test_draw/14_AVT/" + case_name + "_" + str(index) + ".png"
        visualize_segmentation(img, lab, pred,
                               [],
                               [],
                               save_path, "slice_" + str(index) + "_" + case_name, model_name="14_AVT")



import SimpleITK as sitk
from tqdm import tqdm
def test_single_volume_AVT(image, label, net, classes, patch_size=[224, 224], test_save_path=None, case=None,
                           origin = None, direction = None, xyz_thickness = None):
    image, label = image.squeeze(0).cpu().detach().numpy(), label.squeeze(0).cpu().detach().numpy()
    if len(image.shape) == 3:  #3D
        prediction = np.zeros_like(label)
        for ind in tqdm(range(image.shape[0])):
            slice = image[ind, :, :]
            x, y = slice.shape[0], slice.shape[1]
            if x != patch_size[0] or y != patch_size[1]:
                slice = zoom(slice, (patch_size[0] / x, patch_size[1] / y), order=3)  # previous using 0
            input = torch.from_numpy(slice).unsqueeze(0).unsqueeze(0).float().cuda()
            net.eval()
            with torch.no_grad():
                outputs = net(input)
                outputs = outputs[0]
                out = torch.argmax(torch.softmax(outputs, dim=1), dim=1).squeeze(0)
                out = out.cpu().detach().numpy()
                if x != patch_size[0] or y != patch_size[1]:
                    pred = zoom(out, (x / patch_size[0], y / patch_size[1]), order=0)
                else:
                    pred = out
                prediction[ind] = pred
    else:
        input = torch.from_numpy(image).unsqueeze(
            0).unsqueeze(0).float().cuda()
        net.eval()
        with torch.no_grad():
            out = torch.argmax(torch.softmax(net(input), dim=1), dim=1).squeeze(0)
            prediction = out.cpu().detach().numpy()

    draw_slice(net, image, label, prediction, case, patch_size)
    return 0
    # metric_list = []
    # for i in range(1, classes):
    #     metric_list.append(calculate_metric_percase(prediction == i, label == i))
    # 
    # if test_save_path is not None:
    #     img_itk = sitk.GetImageFromArray(image.astype(np.float32))
    #     img_itk.SetOrigin(origin)
    #     img_itk.SetDirection(direction)
    #     img_itk.SetSpacing((xyz_thickness[0], xyz_thickness[1], xyz_thickness[2]))
    #     sitk.WriteImage(img_itk, test_save_path + '/' + case + "_img.nii.gz")
    # 
    #     prd_itk = sitk.GetImageFromArray(prediction.astype(np.float32))
    #     prd_itk.SetOrigin(origin)
    #     prd_itk.SetDirection(direction)
    #     prd_itk.SetSpacing((xyz_thickness[0], xyz_thickness[1], xyz_thickness[2]))
    #     sitk.WriteImage(prd_itk, test_save_path + '/' + case + "_pred.nii.gz")
    # 
    #     lab_itk = sitk.GetImageFromArray(label.astype(np.float32))
    #     lab_itk.SetOrigin(origin)
    #     lab_itk.SetDirection(direction)
    #     lab_itk.SetSpacing((xyz_thickness[0], xyz_thickness[1], xyz_thickness[2]))
    #     sitk.WriteImage(lab_itk, test_save_path + '/' + case + "_gt.nii.gz")

    # return metric_list&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "HongkunSun/ParaTransCNN"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月21日(星期三) 下午5:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [HongkunSun/ParaTransCNN] 关于AVT数据集，为什么我可视化出来的数据不是论文中这种血管树？？？？？？ (Issue #13)



 HongkunSun left a comment (HongkunSun/ParaTransCNN#13)
 
可视化的是标签还是模型推理出来的结果？数据不会出错的

**Comment by sansuitaibai** [comment]
可视化使用3D Slicer，这是什么意思啊，要可视化3D的是数据集？，有没有可视化代码，作者大大




spw
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "HongkunSun/ParaTransCNN"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年5月21日(星期三) 下午5:30
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [HongkunSun/ParaTransCNN] 关于AVT数据集，为什么我可视化出来的数据不是论文中这种血管树？？？？？？ (Issue #13)



 HongkunSun left a comment (HongkunSun/ParaTransCNN#13)
 
可视化使用3D Slicer
